pkg/domain/infosync/affinity_manager.go

```diff
@@ -0,0 +1,157 @@
+// Copyright 2025 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package infosync
+
+import (
+	"context"
+	"sync"
+
+	pdhttp "github.com/tikv/pd/client/http"
+)
+
+// AffinityManager manages affinity groups with PD.
+type AffinityManager interface {
+	CreateAffinityGroupsIfNotExists(ctx context.Context, groups map[string][]pdhttp.AffinityGroupKeyRange) error
+	DeleteAffinityGroups(ctx context.Context, ids []string) error
+	GetAffinityGroups(ctx context.Context, ids []string) (map[string]*pdhttp.AffinityGroupState, error)
+}
+
+type pdAffinityManager struct {
+	pdhttp.Client
+}
+
+// CreateAffinityGroupsIfNotExists creates affinity groups in PD.
+// It checks which groups already exist and only creates the ones that don't exist.
+// This makes the operation safe for DDL job retries.
+func (m *pdAffinityManager) CreateAffinityGroupsIfNotExists(ctx context.Context, groups map[string][]pdhttp.AffinityGroupKeyRange) error {
+	if len(groups) == 0 {
+		return nil
+	}
+
+	// Collect group IDs to check
+	groupIDs := make([]string, 0, len(groups))
+	for id := range groups {
+		groupIDs = append(groupIDs, id)
+	}
+
+	// Check which groups already exist
+	existingGroups, err := m.GetAffinityGroups(ctx, groupIDs)
```

Comment:
I think in the further PRs, PD should provide an idempotent API to batch update / create groups to avoid this get.